packages/realm-server/tests/billing-test.ts

```diff
@@ -0,0 +1,367 @@
+    `INSERT INTO plans (name, monthly_price, credits_included, stripe_plan_id) VALUES`,
+    ...addExplicitParens(separatedByCommas(valueExpressions)),
+    ` RETURNING *`,
+  ] as Expression);
+
+  return {
+    id: result[0].id as string,
+    name: result[0].name as string,
+    monthlyPrice: result[0].monthly_price as number,
+    creditsIncluded: result[0].credits_included as number,
+  };
+}
+
+async function fetchStripeEvents(dbAdapter: PgAdapter) {
+  return await query(dbAdapter, [`SELECT * FROM stripe_events`]);
+}
+
+async function fetchSubscriptionsByUserId(
+  dbAdapter: PgAdapter,
+  userId: string,
+): Promise<Subscription[]> {
+  let results = (await query(dbAdapter, [
+    `SELECT * FROM subscriptions WHERE user_id = `,
+    param(userId),
+  ])) as {
+    id: string;
+    user_id: string;
+    plan_id: string;
+    started_at: number;
+    ended_at: number;
+    status: string;
+    stripe_subscription_id: string;
+  }[];
+
+  return results.map((result) => ({
+    id: result.id,
+    userId: result.user_id,
+    planId: result.plan_id,
+    startedAt: result.started_at,
+    status: result.status,
+    stripeSubscriptionId: result.stripe_subscription_id,
+  }));
+}
+
+async function fetchSubscriptionCyclesBySubscriptionId(
+  dbAdapter: PgAdapter,
+  subscriptionId: string,
+): Promise<SubscriptionCycle[]> {
+  let results = await query(dbAdapter, [
+    `SELECT * FROM subscription_cycles WHERE subscription_id = `,
+    param(subscriptionId),
+  ]);
+
+  return results.map((result) => ({
+    id: result.id as string,
+    subscriptionId: result.subscription_id as string,
+    periodStart: parseInt(result.period_start as string),
+    periodEnd: parseInt(result.period_end as string),
+  }));
+}
+
+async function fetchCreditsLedgerByUser(
+  dbAdapter: PgAdapter,
+  userId: string,
+): Promise<LedgerEntry[]> {
+  let results = await query(dbAdapter, [
+    `SELECT * FROM credits_ledger WHERE user_id = `,
+    param(userId),
+  ]);
+
+  return results.map((result) => ({
+    id: result.id as string,
+    userId: result.user_id as string,
+    creditAmount: result.credit_amount as number,
+    creditType: result.credit_type as string,
+    subscriptionCycleId: result.subscription_cycle_id as string,
+  }));
+}
+
+module('billing', function (hooks) {
+  let dbAdapter: PgAdapter;
+
+  hooks.beforeEach(async function () {
+    prepareTestDB();
+    dbAdapter = new PgAdapter();
+  });
+
+  hooks.afterEach(async function () {
+    await dbAdapter.close();
+  });
+
+  module('invoice payment succeeded', function () {
+    module('new subscription without any previous subscription', function () {
+      test('creates a new subscription and adds plan allowance in credits', async function (assert) {
+        let user = await insertUser(dbAdapter, 'user@test', 'cus_123');
+        let plan = await insertPlan(dbAdapter, 'Free plan', 0, 100, 'prod_123');
+
+        // Omitted version of a real stripe invoice.payment_succeeded event
```

Comment:
What does `Omitted` mean here?

Reply:
The actual event contains many more fields, but our type defines only the ones we care about 

Here is an example of a full event:

```
{
  "object": {
    "id": "[REDACTED_INVOICE_ID]",
    "object": "invoice",
    "account_country": "US",
    "account_name": "[REDACTED_COMPANY_NAME]",
    "account_tax_ids": null,
    "amount_due": 0,
    "amount_paid": 0,
    "amount_remaining": 0,
    "amount_shipping": 0,
    "application": null,
    "application_fee_amount": null,
    "attempt_count": 0,
    "attempted": true,
    "auto_advance": false,
    "automatic_tax": {
      "enabled": true,
      "liability": {
        "type": "self"
      },
      "status": "complete"
    },
    "automatically_finalizes_at": null,
    "billing_reason": "subscription_create",
    "charge": null,
    "collection_method": "charge_automatically",
    "created": "[REDACTED_TIMESTAMP]",
    "currency": "usd",
    "custom_fields": null,
    "customer": "[REDACTED_CUSTOMER_ID]",
    "customer_address": {
      "city": null,
      "country": "[REDACTED]",
      "line1": null,
      "line2": null,
      "postal_code": null,
      "state": null
    },
    "customer_email": "[REDACTED_EMAIL]",
    "customer_name": "[REDACTED]",
    "customer_phone": null,
    "customer_shipping": null,
    "customer_tax_exempt": "none",
    "customer_tax_ids": [],
    "default_payment_method": null,
    "default_source": null,
    "default_tax_rates": [],
    "description": null,
    "discount": null,
    "discounts": [],
    "due_date": null,
    "effective_at": "[REDACTED_TIMESTAMP]",
    "ending_balance": 0,
    "footer": null,
    "from_invoice": null,
    "hosted_invoice_url": "[REDACTED_URL]",
    "invoice_pdf": "[REDACTED_URL]",
    "issuer": {
      "type": "self"
    },
    "last_finalization_error": null,
    "latest_revision": null,
    "lines": {
      "object": "list",
      "data": [{
        "id": "[REDACTED_LINE_ITEM_ID]",
        "object": "line_item",
        "amount": 0,
        "amount_excluding_tax": 0,
        "currency": "usd",
        "description": "1 × Free (at $0.00 / month)",
        "discount_amounts": [],
        "discountable": true,
        "discounts": [],
        "invoice": "[REDACTED_INVOICE_ID]",
        "livemode": false,
        "metadata": {},
        "period": {
          "end": "[REDACTED_TIMESTAMP]",
          "start": "[REDACTED_TIMESTAMP]"
        },
        "plan": {
          "id": "[REDACTED_PLAN_ID]",
          "object": "plan",
          "active": true,
          "aggregate_usage": null,
          "amount": 0,
          "amount_decimal": "0",
          "billing_scheme": "per_unit",
          "created": "[REDACTED_TIMESTAMP]",
          "currency": "usd",
          "interval": "month",
          "interval_count": 1,
          "livemode": false,
          "metadata": {},
          "meter": null,
          "nickname": null,
          "product": "[REDACTED_PRODUCT_ID]",
          "tiers_mode": null,
          "transform_usage": null,
          "trial_period_days": null,
          "usage_type": "licensed"
        },
        "pretax_credit_amounts": [],
        "price": {
          "id": "[REDACTED_PRICE_ID]",
          "object": "price",
          "active": true,
          "billing_scheme": "per_unit",
          "created": "[REDACTED_TIMESTAMP]",
          "currency": "usd",
          "custom_unit_amount": null,
          "livemode": false,
          "lookup_key": null,
          "metadata": {},
          "nickname": null,
          "product": "[REDACTED_PRODUCT_ID]",
          "recurring": {
            "aggregate_usage": null,
            "interval": "month",
            "interval_count": 1,
            "meter": null,
            "trial_period_days": null,
            "usage_type": "licensed"
          },
          "tax_behavior": "unspecified",
          "tiers_mode": null,
          "transform_quantity": null,
          "type": "recurring",
          "unit_amount": 0,
          "unit_amount_decimal": "0"
        },
        "proration": false,
        "proration_details": {
          "credited_items": null
        },
        "quantity": 1,
        "subscription": "[REDACTED_SUBSCRIPTION_ID]",
        "subscription_item": "[REDACTED_SUBSCRIPTION_ITEM_ID]",
        "tax_amounts": [{
          "amount": 0,
          "inclusive": false,
          "tax_rate": "[REDACTED_TAX_RATE_ID]",
          "taxability_reason": "not_collecting",
          "taxable_amount": 0
        }],
        "tax_rates": [],
        "type": "subscription",
        "unit_amount_excluding_tax": "0"
      }],
      "has_more": false,
      "total_count": 1,
      "url": "/v1/invoices/[REDACTED_INVOICE_ID]/lines"
    },
    "livemode": false,
    "metadata": {},
    "next_payment_attempt": null,
    "number": "[REDACTED_INVOICE_NUMBER]",
    "on_behalf_of": null,
    "paid": true,
    "paid_out_of_band": false,
    "payment_intent": null,
    "payment_settings": {
      "default_mandate": null,
      "payment_method_options": {
        "acss_debit": null,
        "bancontact": null,
        "card": {
          "request_three_d_secure": "automatic"
        },
        "customer_balance": null,
        "konbini": null,
        "sepa_debit": null,
        "us_bank_account": null
      },
      "payment_method_types": null
    },
    "period_end": "[REDACTED_TIMESTAMP]",
    "period_start": "[REDACTED_TIMESTAMP]",
    "post_payment_credit_notes_amount": 0,
    "pre_payment_credit_notes_amount": 0,
    "quote": null,
    "receipt_number": null,
    "rendering": null,
    "shipping_cost": null,
    "shipping_details": null,
    "starting_balance": 0,
    "statement_descriptor": null,
    "status": "paid",
    "status_transitions": {
      "finalized_at": "[REDACTED_TIMESTAMP]",
      "marked_uncollectible_at": null,
      "paid_at": "[REDACTED_TIMESTAMP]",
      "voided_at": null
    },
    "subscription": "[REDACTED_SUBSCRIPTION_ID]",
    "subscription_details": {
      "metadata": {}
    },
    "subtotal": 0,
    "subtotal_excluding_tax": 0,
    "tax": 0,
    "test_clock": null,
    "total": 0,
    "total_discount_amounts": [],
    "total_excluding_tax": 0,
    "total_pretax_credit_amounts": [],
    "total_tax_amounts": [{
      "amount": 0,
      "inclusive": false,
      "tax_rate": "[REDACTED_TAX_RATE_ID]",
      "taxability_reason": "not_collecting",
      "taxable_amount": 0
    }],
    "transfer_data": null,
    "webhooks_delivered_at": null
  },
  "previous_attributes": null
}
```